network-config-analyzer/PortSet.py

```diff
@@ -6,294 +6,231 @@
 
-    def __isub__(self, other):
-        self.port_set -= other.port_set
-        self.named_ports -= other.named_ports
-        self.excluded_named_ports |= other.named_ports
-        return self
+class PortSetPair(CanonicalHyperCubeSet):
+    """
+    A class for holding a set of rectangles, each defined over a range of source ports X a range of dest ports
+    All ports are divided to 3 domains:
+    - MIN_PORT_NUM .. MAX_PORT_NUM: a domain for real ports (provided as numbers in policies),
+    - MIN_POS_NAMED_PORT_NUM .. MAX_POS_NAMED_PORT_NUM: a domain for positive named ports (named ports provided under 'ports:' in policies)
+    - MIN_NEG_NAMED_PORT_NUM .. MAX_NEG_NAMED_PORT_NUM: a domain for negative named ports (named ports provided under 'notPorts:' in policies)
 
+    Note, only destination ports may contain named ports.
+    This class makes the following assumption (embodied in assert statements below):
+    since for each rule convert_named_ports is applied, no named ports exist at the moment when logical operations are applied
+    between PortSetPairs (i.e., Allow/Deny logic of rules is executed after all named ports are converted to real ones in the rules)
 
-class PortSetPair:
-    """
-    A class for holding a set of rectangles, each defined over a range of source ports X a range of target ports
     """
 
     def __init__(self, source_ports=PortSet(), dest_ports=PortSet()):
         """
         This will create all rectangles made of a range in source_ports and a range in dest_ports
         :param PortSet source_ports: The set of source ports (as a set of intervals/ranges)
         :param PortSet dest_ports: The set of target ports (as a set of intervals/ranges)
+        assuming named ports are only in dest, not src
         """
-        self.rectangles = CanonicalHyperCubeSet(2)
-        for src in source_ports.port_set:
-            for dst in dest_ports.port_set:
-                rectangle_intervals = [src, dst]
-                self.rectangles.add_interval(rectangle_intervals)
-        self.named_ports = {}  # a mapping from dst named port (String) to src ports interval set
-        self.excluded_named_ports = {}  # a mapping from dst named port (String) to src ports interval set
-        # assuming named ports are only in dest, not src
-        all_ports = CanonicalIntervalSet()
-        all_ports.add_interval(CanonicalIntervalSet.Interval(1, 65536))
-        for port_name in dest_ports.named_ports:
-            self.named_ports[port_name] = source_ports.port_set
-        for port_name in dest_ports.excluded_named_ports:
-            # self.excluded_named_ports[port_name] = all_ports - source_ports.port_set
-            self.excluded_named_ports[port_name] = all_ports
-
-    def __bool__(self):
-        return bool(self.rectangles) or bool(self.named_ports)
+        super().__init__(2)
+        for src in source_ports:
+            for dst in dest_ports:
+                rectangle_intervals = [dst, src]
```

Comment:
the order has changed from (src, dst) to (dst, src ): - should probably change the c'tor args order + in the new branch added dimensions order list -> should update it there as well..